keps/sig-architecture/20191001-prevent-permabeta.md

```diff
@@ -0,0 +1,140 @@
+1. meeting GA criteria and getting promoted, or
+2. having a new beta and deprecating the previous beta
+  
+This must happen within six months (two releases).  If it does not,
+the API will be deprecated with an announced intent to remove the feature per the deprecation policy.
+
+## Motivation
+
+When a feature reaches beta, it is turned on by default.  This is great for getting feedback, but it can also lead to state
+where users and vendors start building important infrastructure against APIs that are not considered stable.
+In addition, once a feature is on by default, the incentive to further stabilize appears to diminish.
+See the features that have been beta for a long time: CSRs and Ingresses as examples.
+If we're honest with ourselves, a single actor has been cleaning up behind a lot of the project to unstick perma-beta APIs.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+
+### Goals
+
+1. Prevent APIs and features from being in a single beta version for more than six months.
+2. Prevent beta APIs from being treated as GA by users and vendors.
+
+### Non-Goals
+
+1. Promote APIs to GA before they are ready.
+
+## Proposal
+
+Once an API or feature reaches beta, it has six months to 
+1. reach GA and deprecate the beta or 
+2. have a new beta version and deprecate the previous beta.
+
+If neither of those conditions met, the beta API/feature is deprecated in the second release with a stated intent to remove the feature entirely.
+To avoid removal, the feature must create a new beta version (it cannot go directly from deprecated to GA).
+
+For example, in v1.16, v1beta1 is released. Sample release note and API doc:
+> * "The v1beta1 version of this API will be evaluated during v1.16 and v1.17, then deprecated in v1.18 (in favor of a new beta version, a GA version, or with no replacement), then removed in v1.21"
+
+Scenario A - progression to v1beta2 in v1.18. Sample release note and API doc:
+> * "The v1beta1 version of this API is deprecated in favor of v1beta2, and will be removed in v1.21"
+> * "The v1beta2 version of this API will be evaluated during v1.18 and v1.19, then deprecated in v1.20 (in favor of a new beta version, a GA version, or with no replacement), then removed in v1.23"
+
+Scenario B - progression to v1 in v1.18. Sample release note and API doc:
+> * "The v1beta1 version of this API is deprecated in favor of v1, and will be removed in v1.21"
+
+Scenario C - deprecation with no replacement. Sample release note and API doc:
+> * "The v1beta1 version of this API is deprecated with no replacement, and will be removed in v1.21"
+
+By regularly having new beta versions, we can ensure that consumers will not grow long running dependencies on particular betas which could pin design decisions.
+It will also create an incentive for feature authors to push their features to GA instead of letting them live in a permanent beta state. 
```

Comment:
The cost to users and vendors of such otherwise unnecessary version churn is out of proportion to their ability to effect the desired outcome.

Treating the users and vendors as a kind of whipping boy to incent the Kubernetes developers is a bit indirect and is unlikely to be terribly effective. Much more effective would be to apply the cost to the Kubernetes developers directly.

Reply:
Ok, I see your point. Your concern is that rather than graduating these beta features to GA, the community will either bump their beta version number or will deprecate them, and either of those is painful for the user community. The concern makes sense to me, but it's not how I see this likely playing out. Here's why.

In the status quo, there is little incentive for developers to move a feature to GA. There is a definite cost to moving to GA, it takes real work and effort. In most environments the feature will be available by default at zero cost by leaving it in beta. I don't think folks are acting in bad faith, it's just that there is always other pressing work, and it's difficult to prioritize something that has cost but little practical value.

The "not acting in bad faith" point is important. To me, the reason this KEP can work without the issues you describe is that I truly believe our community is generally acting in good faith. So, the scenario of perpetual version churn, or the deprecation or removal of beta features, is highly unlikely. Instead, developers will have a good reason to prioritize work on making something GA, rather than moving on to the next feature.

In fact, I think that this is semi-empirical. We can go through the list of current beta features, and reach out to each SIG to get a commitment on the disposition of each of those features should this KEP go into effect. That would give us a preview of the current consequences, and build confidence in our guess at the future consequences.

Reply:
If it were the case that the developers would always graduate the features to GA rather than merely bump the version number while prioritizing other efforts, then there would be no need for this KEP to have a fixed time for a forced deprecation.

For CSR to have not lingered in beta as it did, sig-auth would have had to have not performed one or more of the accomplishments they made during that time and spent the effort put into those accomplishments instead on fixing the numerous deficiencies of CSR. Can we agree that is the goal of this KEP?

Even were this KEP in effect at the time, I cannot say that sig-auth’s prioritizing of the accomplishments they made over fixing CSR would rise to the level of “acting in bad faith”.

The behavior modification technique of applying a cost to a third party so that the second party exhibits the desired behavior through feeling empathy is the classic definition of a “whipping boy”.

Reply:
An important goal of this KEP is to ensure that APIs that have not met the bar for GA (and which are permitted to be removed with as little as 6 months warning) do not get entrenched by large amounts of usage (often in places out of the direct control of the people affected when the beta APIs are finally graduated/deprecated/removed).

Reply:
> If it were the case that the developers would always graduate the features to GA rather than merely bump the version number while prioritizing other efforts, then there would be no need for this KEP to have a fixed time for a forced deprecation.

The forced deprecation is to handle ones that are abandon and never should go to GA.

> The behavior modification technique of applying a cost to a third party so that the second party exhibits the desired behavior through feeling empathy is the classic definition of a “whipping boy”.

Ok, but this does apply a cost to the developers as well. If they have to take some action, it becomes more appealing to just do the right thing and bump it to GA (or eliminate a bad API/feature). I do agree that keeping a limited number in beta as you suggest would help with this too.
 
>An important goal of this KEP is to ensure that APIs that have not met the bar for GA (and which are permitted to be removed with as little as 6 months warning) do not get entrenched by large amounts of usage (often in places out of the direct control of the people affected when the beta APIs are finally graduated/deprecated/removed).

I think this lines right up with the issue @thockin raised - how do we make "beta" more visible for the end user, so they are more cognizant of using a non-stable API? This would be a complementary, separate effort. One way is to not build those in - but then we don't get the usage information. Another is to hide the beta endpoints and fields unless you apply a special header (added with a "beta" flag when using `kubectl` for example). The "beta" in the kind doesn't seem to be enough.